File: BrewPOS-Workspace/brewpos-backend/src/services/gamificationService.ts
import { prisma } from '../db';

/**
 * Memproses Gamifikasi (Quest & Badges) secara asynchronous.
 * Dipanggil setelah proses checkout selesai agar tidak memblokir response ke kasir.
 */
export async function processGamificationAsync(customerId: number, transactionId: number) {
  try {
    const transaction = await prisma.transaction.findUnique({
      where: { id: transactionId },
      include: { items: true },
    });

    if (!transaction) return;

    // --- 1. EVALUASI QUESTS ---
    const quests = await prisma.quest.findMany({ where: { active: true } });
    
    for (const quest of quests) {
      let customerQuest = await prisma.customerQuest.findUnique({
        where: { customerId_questId: { customerId, questId: quest.id } }
      });

      if (!customerQuest) {
        customerQuest = await prisma.customerQuest.create({
          data: { customerId, questId: quest.id }
        });
      }

      if (!customerQuest.isCompleted) {
        let addedProgress = 0;

        if (quest.type === 'TOTAL_TRANSACTIONS') {
          addedProgress = 1;
        } else if (quest.type === 'TOTAL_SPEND') {
          addedProgress = transaction.totalAmount;
        } else if (quest.type === 'BUY_ITEM' && quest.targetEntityId) {
          const matchingItems = transaction.items.filter((item: any) => item.menuId === quest.targetEntityId);
          addedProgress = matchingItems.reduce((acc: number, curr: any) => acc + curr.quantity, 0);
        }

        if (addedProgress > 0) {
          const newProgress = customerQuest.progress + addedProgress;
          const isCompleted = newProgress >= quest.targetValue;

          await prisma.customerQuest.update({
            where: { id: customerQuest.id },
            data: { progress: newProgress, isCompleted }
          });

          if (isCompleted) {
            // Berikan reward karena quest selesai
            await prisma.customer.update({
              where: { id: customerId },
              data: {
                points: { increment: quest.rewardPoints },
                xp: { increment: quest.rewardXp }
              }
            });
            console.log(`[Gamification] Customer ${customerId} completed quest: ${quest.name}`);
          }
        }
      }
    }

    // --- 2. EVALUASI BADGES ---
    const badges = await prisma.badge.findMany();
    const count = await prisma.transaction.count({ where: { customerId } });

    for (const badge of badges) {
      const hasBadge = await prisma.customerBadge.findUnique({
        where: { customerId_badgeId: { customerId, badgeId: badge.id } }
      });
      
      if (!hasBadge) {
        if (badge.requiredTransactions > 0 && count >= badge.requiredTransactions) {
          await prisma.customerBadge.create({
            data: { customerId, badgeId: badge.id }
          });
          console.log(`[Gamification] Customer ${customerId} unlocked badge: ${badge.name}`);
        }
      }
    }

  } catch (error) {
    console.error("[Gamification] Error processing gamification asynchronously:", error);
  }
}

File: BrewPOS-Workspace/brewpos-flask/app/services/gamification_service.py
import threading
from flask import current_app
from app.extensions import db
from app.models.customer import Customer, CustomerQuest, CustomerBadge
from app.models.quest import Quest
from app.models.badge import Badge
from app.models.transaction import Transaction

def _process_gamification(app, customer_id, transaction_id):
    with app.app_context():
        try:
            transaction = Transaction.query.get(transaction_id)
            customer = Customer.query.get(customer_id)
            if not transaction or not customer:
                return

            # --- 1. EVALUATE QUESTS ---
            active_quests = Quest.query.filter_by(active=True).all()
            for quest in active_quests:
                cq = CustomerQuest.query.filter_by(customerId=customer_id, questId=quest.id).first()
                if not cq:
                    cq = CustomerQuest(customerId=customer_id, questId=quest.id, progress=0, isCompleted=False)
                    db.session.add(cq)
                    db.session.flush()

                if not cq.isCompleted:
                    added_progress = 0

                    if quest.type == 'TOTAL_TRANSACTIONS':
                        added_progress = 1
                    elif quest.type == 'TOTAL_SPEND':
                        added_progress = transaction.totalAmount
                    elif quest.type == 'BUY_ITEM' and quest.targetEntityId:
                        matching_items = [it for it in transaction.items if it.menuId == quest.targetEntityId]
                        added_progress = sum(it.quantity for it in matching_items)

                    if added_progress > 0:
                        cq.progress += added_progress
                        if cq.progress >= quest.targetValue:
                            cq.isCompleted = True
                            customer.points += quest.rewardPoints
                            customer.xp += quest.rewardXp
                            print(f"[Gamification] Customer {customer.nickname} completed quest '{quest.name}'")

            # --- 2. EVALUATE BADGES ---
            badges = Badge.query.all()
            total_tx_count = Transaction.query.filter_by(customerId=customer_id, status='COMPLETED').count()

            for badge in badges:
                has_badge = CustomerBadge.query.filter_by(customerId=customer_id, badgeId=badge.id).first()
                if not has_badge:
                    if badge.requiredTransactions > 0 and total_tx_count >= badge.requiredTransactions:
                        cb = CustomerBadge(customerId=customer_id, badgeId=badge.id)
                        db.session.add(cb)
                        print(f"[Gamification] Customer {customer.nickname} unlocked badge '{badge.name}'")

            db.session.commit()
        except Exception as e:
            db.session.rollback()
            print(f"[Gamification] Error processing gamification: {e}")

def process_gamification_async(customer_id, transaction_id):
    """
    Spawns a background thread to process quests & badges non-blockingly.
    """
    app = current_app._get_current_object()
    thread = threading.Thread(target=_process_gamification, args=(app, customer_id, transaction_id))
    thread.daemon = True
    thread.start()
